examples/click_rate_demo/ctr_utils.hpp

```diff
@@ -1,11 +1,47 @@
 #pragma once
 #include <memory>
 #include <string>
+#include <map>
 #include <type_traits>
+
 #include <thrift/transport/TBufferTransports.h>
 #include <thrift/protocol/TBinaryProtocol.h>
 #include <thrift/protocol/TJSONProtocol.h>
 
+#include "gen-cpp/click_types.h"
+
+enum class Publisher {
+  ADSTERRA,
+  UNDERTONE,
+  CLICKSOR,
+  NEXUS,
+  CONVERSANT
+};
+
+std::map<std::string, Publisher> string_to_publisher = {
+  { "ADSTERRA", Publisher::ADSTERRA },
+  { "UNDERTONE", Publisher::UNDERTONE },
+  { "CLICKSOR", Publisher::CLICKSOR },
+  { "NEXUS", Publisher::NEXUS },
+  { "CONVERSANT", Publisher::CONVERSANT }
+};
+
+thrift::AdEvent newEvent(const thrift::StreamEvent::type evnt,
+                         const int32_t eId) {
+  thrift::AdEvent event;
+  event.__set_type(evnt);
```

Comment:
call the 

`event.type = evnt`

directly. 

.__set_type is not needed because you made them required.